ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java

```diff
@@ -0,0 +1,239 @@
+import org.opensearch.ml.common.input.Input;
+import org.opensearch.ml.common.input.execute.agent.AgentMLInput;
+import org.opensearch.ml.common.output.Output;
+import org.opensearch.ml.common.output.model.ModelTensor;
+import org.opensearch.ml.common.output.model.ModelTensorOutput;
+import org.opensearch.ml.common.output.model.ModelTensors;
+import org.opensearch.ml.common.spi.memory.Memory;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.ml.engine.Executable;
+import org.opensearch.ml.engine.annotation.Function;
+import org.opensearch.ml.engine.memory.ConversationIndexMemory;
+import org.opensearch.ml.engine.memory.ConversationIndexMessage;
+import org.opensearch.ml.memory.action.conversation.CreateInteractionResponse;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Data
+@NoArgsConstructor
+@Function(FunctionName.AGENT)
+public class MLAgentExecutor implements Executable {
+
+    public static final String MEMORY_ID = "memory_id";
+    public static final String QUESTION = "question";
+    public static final String PARENT_INTERACTION_ID = "parent_interaction_id";
+
+    private Client client;
+    private Settings settings;
+    private ClusterService clusterService;
+    private NamedXContentRegistry xContentRegistry;
+    private Map<String, Tool.Factory> toolFactories;
+    private Map<String, Memory.Factory> memoryFactoryMap;
+
+    public MLAgentExecutor(
+        Client client,
+        Settings settings,
+        ClusterService clusterService,
+        NamedXContentRegistry xContentRegistry,
+        Map<String, Tool.Factory> toolFactories,
+        Map<String, Memory.Factory> memoryFactoryMap
+    ) {
+        this.client = client;
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.xContentRegistry = xContentRegistry;
+        this.toolFactories = toolFactories;
+        this.memoryFactoryMap = memoryFactoryMap;
+    }
+
+    @Override
+    public void execute(Input input, ActionListener<Output> listener) {
+        if (!(input instanceof AgentMLInput)) {
+            throw new IllegalArgumentException("wrong input");
+        }
+        AgentMLInput agentMLInput = (AgentMLInput) input;
+        String agentId = agentMLInput.getAgentId();
+        RemoteInferenceInputDataSet inputDataSet = (RemoteInferenceInputDataSet) agentMLInput.getInputDataset();
+        if (inputDataSet.getParameters() == null) {
+            throw new IllegalArgumentException("wrong input");
+        }
+
+        List<ModelTensors> outputs = new ArrayList<>();
+        List<ModelTensor> modelTensors = new ArrayList<>();
+        outputs.add(ModelTensors.builder().mlModelTensors(modelTensors).build());
+
+        if (clusterService.state().metadata().hasIndex(ML_AGENT_INDEX)) {
```

Comment:
this should be something like
```java
        if (!clusterService.state().metadata().hasIndex(ML_AGENT_INDEX)) {
            listener.onFailure(new ResourceNotFoundException("Agent index not found"));
        }
```
it's not returning a response when index is not found

Reply:
Thanks Joshua. @jngz-es help fix this part?

Reply:
sure, will fix it in my pr.

Reply:
Fixed in https://github.com/opensearch-project/ml-commons/pull/1866